### ISSUE TITLE
Go To Type Definition on arrays

### ISSUE THREAD

**Issue by normalser** [microsoft/TypeScript#14148]
It would be great if `Go To Type Definition` used on element that is array of some interface -  would go to definition of this interface. Right now it goes to `Array interface` from `lib.es5.d.ts` - which in theory one could say 'Working as intended' but in practice there is not a lot of value to go to definition of `Array` each time

```ts
interface Test {

}

function foo() {
    const a:Test = {}
    return [a]
}

const goToTypeDefinitionOnMe = foo()
```

EDIT: Eventually if Array is detected then new entry `Go To Type Definition of child`

**Comment by mhegazy** [comment]
PRs welcomed. should be a fairly small change in [services/gotoDefinition.ts](https://github.com/Microsoft/TypeScript/blob/28041547784b5c5d73b50472bbafe39137fed7dd/src/services/goToDefinition.ts#L139)

**Comment by DanielRosenwasser** [comment]
My concern about this is that it wouldn't also work for `Iterable`, `Promise`, and Observable`.

**Comment by andrewbranch** [comment]
Assigning to myself because this has annoyed me several times and I should have time to pick it up for 3.8, but anyone is still free to take it until I remove “help wanted.”

**Comment by andrewbranch** [comment]
> As a note, my concern about this is that it wouldn't also work for Iterable, Promise, and Observable.

My thought was that it should return the element type definition _and_ the container type definition (in that order? in the reverse order?), and there should either be a list of built-in container types that this behavior applies to (`Array`, `ReadonlyArray`, `Iterable`, `Promise`, `Set`, etc.), or maybe even automatically do this for any generic type with one type parameter? That way

```ts
interface Box<T> { /* ... */ }
interface Cat { /* ... */ }

declare let thoughtExperiment: Box<Cat>;
//          ^ Go to Type Definition:
//            ---------------------
//            1. interface Box<T> {
//            2. interface Cat {
```

It would be better if the definitions UI let us make labeled groups (this would be useful for other features we’ve added too, like low-confidence “unverified” definitions. But this might be ok to start with?
